### PR TITLE
chore+fix: Cleanup and minor bugfix in `select_by_mask()` methods

### PR DESCRIPTION
**Pull request by vinamarora8**
**Main changes:**
- Removes `**kwargs` as an input parameter to `ArrayDict.select_by_mask()`. This was a TODO item in the code.
- Now all private attributes are deepcopied (this was a bug in Lazy objects)
- Adjusts `IrregularTS` and `Interval` to account for their special private attribs (`_sorted`, `_domain`, and `_timekeys`)
- Input validation checks converted from assertsions to `ValueError` + they are now refactored in a single function.
- Adds missing docstrings for `*.select_by_mask()`

**Additional change that is somewhat related:**
Also removes class-level mutable defaults for `_lazy_ops/_unicode_keys` on Lazy* (shared-class-state potential issue surfaced during review of select_by_mask by copilot (l[ink](https://github.com/neuro-galaxy/temporaldata/pull/121#discussion_r3270871131))). I have explicitly forbid direct construction of Lazy* objects, which was always the case implicitly.

**Tests:**
- Input validation tests
- Tests to ensure private attribs are maintained (some of these would fail on current `main` branch)
- Tests to ensure that things that should be deepcopied are actually deepcopied

**TODO:** 
- [x] Add tests for mask dimension mismatch for IrregularTS and ArrayDict
- [x] Add tests for `_sorted` handling (unsorted things should be masked into sorted=None)
- [x] IrregularTS: Add tests for domain being copied correctly (not a ref)

### CODE REVIEW

File: temporaldata/arraydict.py [hunk 1]
```diff
@@ -129,21 +127,27 @@ def select_by_mask(self, mask: np.ndarray, **kwargs):
...
-        # kwargs are other private attributes
-        # TODO automatically add private attributes
-        return self.__class__(
-            **{k: getattr(self, k)[mask].copy() for k in self.keys()}, **kwargs
-        )
+        new_data = {
+            k: (
+                self.__dict__[k][mask].copy()
+                if not k.startswith("_")
+                else copy.copy(self.__dict__[k])
```
Comment (AlexandreAndr) [inline temporaldata/arraydict.py]: why do we need a different call of copy here (i.e., why not `self.__dict__[k].copy()`)

File: temporaldata/arraydict.py [hunk 2]
```diff
@@ -394,39 +398,36 @@ def __getattribute__(self, name):
...
-        if "mask" not in self._lazy_ops:
+        # combine mask with any pre-existing lazy mask
+        out._lazy_ops = copy.copy(self._lazy_ops)
+        if "mask" not in out._lazy_ops:
             out._lazy_ops["mask"] = mask
         else:
-            # if a mask was already applied, we need to combine the masks
-            out._lazy_ops["mask"] = self._lazy_ops["mask"].copy()
+            out._lazy_ops["mask"] = out._lazy_ops["mask"].copy()
             out._lazy_ops["mask"][out._lazy_ops["mask"]] = mask
```
[the remark this answers is not in the document]
Comment (AlexandreAndr) [reply temporaldata/arraydict.py:426]: up on this

Comment (AlexandreAndr) [inline temporaldata/arraydict.py:426]: is there the mask copy issue here too?

File: tests/test_select_by_mask.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    tmpfile = tempfile.NamedTemporaryFile(suffix=".h5", delete=False)
+    filepath = tmpfile.name
+
+    def finalizer():
+        tmpfile.close()
+        if os.path.exists(filepath):
+            os.remove(filepath)
+
+    request.addfinalizer(finalizer)
+    return filepath
```
[the remark this answers is not in the document]
Comment (AlexandreAndr) [reply tests/test_select_by_mask.py:32]: up on this

**Review by AlexandreAndr — CHANGES_REQUESTED** [review]
Looks great to me just one question on one copy and the comments from coderabbit

File: temporaldata/arraydict.py (hunk 1 above)
Comment (vinamarora8) [reply temporaldata/arraydict.py]: The first copy call is supposedly on numpy arrays (which have .copy()), and the second call could be in any python object

Comment (vinamarora8) [reply temporaldata/arraydict.py]: I could make both the same i guess

Comment (AlexandreAndr) [reply temporaldata/arraydict.py]: Make sense! As you want I think there is no impact on clarity or anything

File: temporaldata/arraydict.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (vinamarora8) [reply temporaldata/arraydict.py:426]: amazing catch!!

[the remark this answers is not in the document]
Comment (vinamarora8) [reply temporaldata/arraydict.py:426]: fixed

Comment (vinamarora8) [reply temporaldata/arraydict.py:426]: I think this is fine? since 
```python
out._lazy_ops["mask"][out._lazy_ops["mask"]] = mask
```
Should already be copying `mask` into `out._lazy_pos["mask"]`. I will nonetheless add tests for this.

File: tests/test_select_by_mask.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (vinamarora8) [reply tests/test_select_by_mask.py:32]: fixed

**Comment by vinamarora8** [comment]
Changes: 
- Fix mask-reference bug in Lazy* objects. Legit bug. I have added a test that was failing before I fixed the bug.
- Renamed `test_select_by_mask_input_validation` to `test_select_by_mask.py` (we can put more select_by_mask tests in here, since they share similar ideas across objects)

I realized handing `_domain`, `_timekeys` needs some extra work and extra tests. I will take care of this tomorrow.
Notes to self: `_domain` needs a deepcopy, since shallow-copy keeps the start/end numpy array refs 
(`x = Interval(0, 1); id(copy.copy(x).start) == id(x.start)`)

File: tests/test_select_by_mask.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        domain="auto",
+    )
+
+
+def _make_lazy(non_lazy, lazy_cls, test_filepath):
+    with h5py.File(test_filepath, "w") as f:
+        non_lazy.to_hdf5(f)
+    f = h5py.File(test_filepath, "r")
+    return lazy_cls.from_hdf5(f), f
+
```
[the remark this answers is not in the document]
Comment (vinamarora8) [reply tests/test_select_by_mask.py:60]: fixed

File: temporaldata/arraydict.py [hunk 5]
```diff
@@ -394,39 +391,33 @@ def __getattribute__(self, name):
...
+        _validate_select_by_mask_input(mask, len(self))
+
+        out = self.__class__.__new__(self.__class__)
+        for key, value in self.__dict__.items():
+            if key.startswith("_"):
+                out.__dict__[key] = copy.deepcopy(value)
+            elif isinstance(value, h5py.Dataset):
+                # mask will be applied lazily on attribute access via _lazy_ops
                 out.__dict__[key] = value
             else:
```
Comment (AlexandreAndr) [inline temporaldata/arraydict.py]: can we add the np.array case as a elif and have as else a error?

File: temporaldata/interval.py [hunk 6]
```diff
@@ -911,38 +921,37 @@ def __getattribute__(self, name):
...
+        _validate_select_by_mask_input(mask, len(self))
+
+        out = self.__class__.__new__(self.__class__)
+        for key, value in self.__dict__.items():
+            if key.startswith("_"):
+                out.__dict__[key] = copy.deepcopy(value)
+            elif isinstance(value, h5py.Dataset):
+                # mask will be applied lazily on attribute access via _lazy_ops
                 out.__dict__[key] = value
             else:
```
Comment (AlexandreAndr) [inline temporaldata/interval.py]: same comment that before check it is a np.array and else be error

File: temporaldata/irregular_ts.py [hunk 7]
```diff
@@ -509,42 +521,37 @@ def __getattribute__(self, name):
...
+        _validate_select_by_mask_input(mask, len(self))
+
+        out = self.__class__.__new__(self.__class__)
+        for key, value in self.__dict__.items():
+            if key.startswith("_"):
+                out.__dict__[key] = copy.deepcopy(value)
+            elif isinstance(value, h5py.Dataset):
+                # mask will be applied lazily on attribute access via _lazy_ops
                 out.__dict__[key] = value
             else:
```
Comment (AlexandreAndr) [inline temporaldata/irregular_ts.py]: same as above

File: temporaldata/regular_ts.py [hunk 8]
```diff
@@ -86,6 +86,13 @@ def domain(self) -> Interval:
...
 
     def select_by_mask(self, mask: np.ndarray):
+        """Raises a NotImplementedError as this method is not supported
+        for :obj:`RegularTimeSeries`.
+
+        Raises:
+            NotImplementedError: Always, because this method cannot
+                be implemented for this class.
+        """
         raise NotImplementedError("Not implemented for RegularTimeSeries.")
```
Comment (AlexandreAndr) [inline temporaldata/regular_ts.py:97]: do we want to add a todo to say that could be implemented when gappy regular ts exists?

File: tests/test_interval.py [hunk 9]
```diff
@@ -247,6 +247,45 @@ def test_lazy_interval(test_filepath):
         assert np.allclose(data.start, np.array([0, 3]))
 
 
+def test_interval_select_by_mask_preserves_private_attrs():
+    data = Interval(
```
Comment (AlexandreAndr) [inline tests/test_interval.py:251]: is there a benefit to have this as a fixture or function to not duplicate data creation between the two tests?

File: tests/test_irregular_ts.py [hunk 10]
```diff
@@ -324,6 +324,52 @@ def test_lazy_irregular_timeseries(test_filepath):
         assert np.allclose(data.values, np.array([1, 3]))
 
 
+def test_irregular_select_by_mask_preserves_private_attrs():
+    data = IrregularTimeSeries(
```
Comment (AlexandreAndr) [inline tests/test_irregular_ts.py:328]: is there a benefit to have this as a fixture or function to not duplicate data creation between the two tests?

File: tests/test_irregular_ts.py [hunk 11]
```diff
@@ -324,6 +324,52 @@ def test_lazy_irregular_timeseries(test_filepath):
...
+    assert result.timekeys() == ["timestamps", "go_cue_time"]
+    assert result._sorted is True
+    assert np.array_equal(result.domain.start, data.domain.start)
+    assert np.array_equal(result.domain.end, data.domain.end)
+
+
+def test_lazy_irregular_select_by_mask_preserves_private_attrs(test_filepath):
+    # `select_by_mask` on a Lazy object must auto-propagate private attrs set
+    # by from_hdf5 (e.g. `_sorted`), not just the hand-listed ones.
+    domain = Interval(start=np.array([0.0]), end=np.array([1.0]))
```
Comment (AlexandreAndr) [inline tests/test_irregular_ts.py]: why domain is def here?

**Review by AlexandreAndr — COMMENTED** [review]
Really cool tests sets and cleanner code in addition to the bug fix only minor comment!

File: tests/test_interval.py (hunk 9 above)
Comment (vinamarora8) [reply tests/test_interval.py:251]: This internval has been duplicated many many times in this test file. I think we should have another PR that cleans up the test files

File: tests/test_irregular_ts.py (hunk 10 above)
Comment (vinamarora8) [reply tests/test_irregular_ts.py:328]: same as previous reply

File: tests/test_irregular_ts.py (hunk 11 above)
Comment (vinamarora8) [reply tests/test_irregular_ts.py]: fixed!

File: tests/test_interval.py (hunk 9 above)
Comment (AlexandreAndr) [reply tests/test_interval.py:251]: Make sense!

**Review by AlexandreAndr — APPROVED** [review]

**Comment by AlexandreAndr** [comment]
LGTM!
